### ISSUE TITLE
[jvm-packages] xgboost4j-spark 0.8: customized objective function "obj_type" error

### ISSUE THREAD

**Issue by iwen-kang** [dmlc/xgboost#3620]
Similar to the example code in https://github.com/dmlc/xgboost/issues/3468, attached please find my testing code:

import ml.dmlc.xgboost4j.scala.{Booster, DMatrix, EvalTrait, ObjectiveTrait}
import ml.dmlc.xgboost4j.scala.spark.XGBoost
import ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier
import org.apache.spark.sql.SparkSession
import org.apache.spark.SparkConf
import java.io.File
import java.io.PrintWriter

import ml.dmlc.xgboost4j.java.XGBoostError
import org.apache.commons.logging.{Log, LogFactory}

import scala.collection.mutable.ListBuffer

//./bin/spark-runner -m cluster -n XgbSparkExample -- 20 2 /user/ekang/libsvm.txt.train /user/ekang/libsvm.txt.test

object XgbSparkExample {

  /**
    * loglikelihoode loss obj function
    */
  class LogRegObj extends ObjectiveTrait {
    private val logger: Log = LogFactory.getLog(classOf[LogRegObj])

    /**
      * user define objective function, return gradient and second order gradient
      *
      * @param predicts untransformed margin predicts
      * @param dtrain   training data
      * @return List with two float array, correspond to first order grad and second order grad
      */
    override def getGradient(predicts: Array[Array[Float]],
                             dtrain: DMatrix): List[Array[Float]] = {
      val nrow = predicts.length
      val gradients = new ListBuffer[Array[Float]]
      var labels: Array[Float] = null
      try {
        labels = dtrain.getLabel
      } catch {
        case e: XGBoostError =>
          logger.error(e)
          null
        case _: Throwable =>
          null
      }
      val grad = new Array[Float](nrow)
      val hess = new Array[Float](nrow)
      val transPredicts = transform(predicts)

      for (i <- 0 until nrow) {
        val predict = transPredicts(i)(0)
        grad(i) = predict - labels(i)
        hess(i) = predict * (1 - predict)
      }
      gradients += grad
      gradients += hess
      gradients.toList
    }

    /**
      * simple sigmoid func
      *
      * @param input
      * @return Note: this func is not concern about numerical stability, only used as example
      */
    def sigmoid(input: Float): Float = {
      (1 / (1 + Math.exp(-input))).toFloat
    }

    def transform(predicts: Array[Array[Float]]): Array[Array[Float]] = {
      val nrow = predicts.length
      val transPredicts = Array.fill[Float](nrow, 1)(0)
      for (i <- 0 until nrow) {
        transPredicts(i)(0) = sigmoid(predicts(i)(0))
      }
      transPredicts
    }

  }

  class EvalError extends EvalTrait {

    val logger = LogFactory.getLog(classOf[EvalError])

    //private[xgboost4j] var evalMetric: String = "custom_error"
    private var evalMetric: String = "custom_error"

    /**
      * get evaluate metric
      *
      * @return evalMetric
      */
    override def getMetric: String = evalMetric

    /**
      * evaluate with predicts and data
      *
      * @param predicts predictions as array
      * @param dmat     data matrix to evaluate
      * @return result of the metric
      */
    override def eval(predicts: Array[Array[Float]], dmat: DMatrix): Float = {
      var error: Float = 0f
      var labels: Array[Float] = null
      try {
        labels = dmat.getLabel
      } catch {
        case ex: XGBoostError =>
          logger.error(ex)
          return -1f
      }
      val nrow: Int = predicts.length
      for (i <- 0 until nrow) {
        if (labels(i) == 0.0 && predicts(i)(0) > 0) {
          error += 1
        } else if (labels(i) == 1.0 && predicts(i)(0) <= 0) {
          error += 1
        }
      }
      error / labels.length
    }
  }

  def saveDumpModel(modelPath: String, modelInfos: Array[String]): Unit = {
    val writer = new PrintWriter(modelPath, "UTF-8")
    for (i <- 0 until modelInfos.length) {
      writer.print(s"booster[$i]:\n")
      writer.print(modelInfos(i))
    }
    writer.close()
  }

  def main(args: Array[String]): Unit = {
    if (args.length != 4) {
      println(
        "usage: program num_of_rounds num_workers training_path test_path")
      sys.exit(1)
    }
    val startTime: Long = System.currentTimeMillis
    // create SparkSession
    val sparkConf = new SparkConf()
      .setAppName("XGBoost-spark-example")
      .set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
    sparkConf.registerKryoClasses(Array(classOf[Booster]))
    // val sqlContext = new SQLContext(new SparkContext(sparkConf))
    val sparkSession = SparkSession.builder().config(sparkConf).getOrCreate()
    val sc = sparkSession.sparkContext
    implicit val sparkContext = sc
    //val sparkSession = SparkSession.builder().config(sparkConf).master("local[2]").getOrCreate()
    // create training and testing dataframes
    val numRound = args(0).toInt
    val numWorkers = args(1).toInt //2
    println("XGBoost numWorkers=" + numWorkers)
    val inputTrainPath = args(2)
    val inputTestPath = args(3)
    // build dataset
    val trainDF =
      sparkSession.sqlContext.read.format("libsvm").load(inputTrainPath)
    //trainDF.repartition(1)
    val testDF =
      sparkSession.sqlContext.read.format("libsvm").load(inputTestPath)

    // start training
    //custom objective no need to set "objective"
    val paramMap = List("eta" -> 0.1f,
                        "max_depth" -> 2,
                        "custom_obj" -> new LogRegObj,
                        "custom_eval" -> new EvalError).toMap
    println("XGBoost paramMap=" + paramMap)
    val xgb = new XGBoostClassifier(paramMap)
    val xgboostModel = xgb.fit(trainDF)
    // xgboost-spark appends the column containing prediction results
    xgboostModel.transform(testDF).show()
    println(
      "xgboostModel transform done! Time spent (in sec): " + ((System.currentTimeMillis - startTime) / 1000)
    )
  }
}

After running it under the latest xgboost4j-spark 0.8 over Spark 2.3.1, I got the “obj_type” error - see below.
User class threw exception: java.lang.IllegalArgumentException: requirement failed: parameter "obj_type" is not defined, you have to specify the objective type as classification or regression with a customized objective function
at scala.Predef$.require(Predef.scala:224)
at ml.dmlc.xgboost4j.scala.spark.XGBoost$.trainDistributed(XGBoost.scala:196)
at ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier.train(XGBoostClassifier.scala:191)
at ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier.train(XGBoostClassifier.scala:48)
at org.apache.spark.ml.Predictor.fit(Predictor.scala:118)
at com.apple.siri.applications.speechData.XgbSparkExample$.main(XgbSparkExample.scala:179)
at com.apple.siri.applications.speechData.XgbSparkExample.main(XgbSparkExample.scala)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$4.run(ApplicationMaster.scala:721)

Looks like some people also ran into same problem and was unable to get a resolution, but it was an old issue posted pre 0.8 release (see https://github.com/dmlc/xgboost/issues/2002) so I wonder if this is fixed in 0.8? If yes, can you please provide instructions to set “obj_type” for customized objective function? Thanks!

Environment:
uname -a
Linux xxxxxx.com 2.6.32-431.17.1.el6.x86_64 #1 SMP Wed May 7 12:11:08 PDT 2014 x86_64 x86_64 x86_64 GNU/Linux

Spark version: 2.3.1
SPARK_ARGS="${SPARK_ARGS} --conf spark.yarn.appMasterEnv.JAVA_HOME=/opt/oracle-jdk-1.8.0_152"

**Comment by CodingCat** [comment]
known issue, fix coming

**Comment by CodingCat** [comment]
@mmui will give a fix soon

**Comment by CodingCat** [comment]
@mmui any update on this?

**Comment by mmui** [comment]
@CodingCat @iwen-kang yep this is due to obj_type not being handled properly in a number of places e.g. LearningTaskParams. Testing my fix right now, creating a PR for this today.

**Comment by mmui** [comment]
https://github.com/dmlc/xgboost/pull/3646
